src/subscription.ts

```diff
@@ -88,6 +88,20 @@ export type SetSubscriptionMetadataResponse = MetadataResponse;
 export type DetachSubscriptionCallback = EmptyCallback;
 export type DetachSubscriptionResponse = EmptyResponse;
 
+// JSDoc won't see these, so this is just to let you get typings
+// in your editor of choice. Callback types are open ended so that
+// not-quite-matching parameters won't cause a breaking change. We
+// may still shift these to required later.
+export declare interface Subscription {
+  on(event: 'message', listener: Function | ((message: Message) => void)): this;
```

Comment:
The use of `Function | <blergh>` to me here is strange, because you're providing a nice function type for the user, and then making it more broadly just "Function".  Can we get rid of the `Function |` piece, and always keep stricter bindings?

Reply:
Mentioned on chat, but bringing it here for reference: without the `Function` it becomes a compile time breaking change for TypeScript users, because they may be passing in valid (but non-matching) callback handlers. For example, you could have a unit test that just passes in `resolve` to `on()`, and that would work before, but may not work now. My intent here is basically to give IDEs a breadcrumb trail to show the user the types, without breaking any existing code.

Reply:
Ah, in that case - I would still drop `Function` (I think) and move to `(message?: Message) => void` so the parameter is optional.  Would that get us where we want to go?

Reply:
As a user I would expect the types to show whats possible in the api, optional when in fact it is not, would just confuse me. If its possible that on('message') will call my handler without a message parameter than the optional should be there to reflect that, on the other hand if there is no such case, then optional would make me think I have to handle a case that is not possible.
e.g: how it is done for node [process](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/process.d.ts#L309) is very clear
Test use-cases I would not consider for the legit types. Promise resolve is not going to break with a message, but in tests also any stub should work

my 2 cents from the user perspective, hope it helps :)